src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs

```diff
@@ -0,0 +1,430 @@
+
+                    case QUIC_CONNECTION_EVENT.SHUTDOWN_BEGIN:
+                        {
+                            status = HandleEventShutdownBegin(
+                                connectionEvent);
+                        }
+                        break;
+
+                    case QUIC_CONNECTION_EVENT.SHUTDOWN_BEGIN_PEER:
+                        {
+                            status = HandleEventShutdownBeginPeer(
+                                connectionEvent);
+                        }
+                        break;
+
+                    case QUIC_CONNECTION_EVENT.SHUTDOWN_COMPLETE:
+                        {
+                            status = HandleEventShutdownComplete(
+                                connectionEvent);
+                        }
+                        break;
+
+                    case QUIC_CONNECTION_EVENT.NEW_STREAM:
+                        {
+                            status = HandleEventNewStream(
+                                connectionEvent);
+                        }
+                        break;
+
+                    case QUIC_CONNECTION_EVENT.STREAMS_AVAILABLE:
+                        {
+                            status = HandleEventStreamsAvailable(
+                                connectionEvent);
+                        }
+                        break;
+
+                    default:
+                        break;
+                }
+            }
+            catch (Exception)
+            {
+                return MsQuicConstants.InternalError;
+            }
+
+            return status;
+        }
+
+        private uint HandleEventConnected(ConnectionEvent connectionEvent)
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            _connectTcs?.SetResult(null);
+            _connectTcs = null;
+
+            if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+            return MsQuicConstants.Success;
+        }
+
+        private uint HandleEventShutdownBegin(ConnectionEvent connectionEvent)
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            if (NetEventSource.IsEnabled) NetEventSource.Info(this, $"Shutdown begin {MsQuicConstants.ErrorTypeFromErrorCode(connectionEvent.ShutdownBeginStatus)}");;
+
+            _connectTcs?.SetResult(null);
+            _connectTcs = null;
+
+            if (NetEventSource.IsEnabled) NetEventSource.Exit(this);
+
+            return MsQuicConstants.Success;
+        }
+
+        private uint HandleEventShutdownBeginPeer(ConnectionEvent connectionEvent)
+        {
+            return MsQuicConstants.Success;
+        }
+
+        private uint HandleEventShutdownComplete(ConnectionEvent connectionEvent)
+        {
+            if (NetEventSource.IsEnabled) NetEventSource.Enter(this);
+
+            if (NetEventSource.IsEnabled) NetEventSource.Info(this, $"Shutdown Complete");
+
+            _shutdownTcs?.SetResult(null);
```

Comment:
Does this mean `_shutdownTcs` can be `null`? How does that affect the shutdown function below?